### PR TITLE
[Feaute]support spconv2.0

### PR DESCRIPTION
**Pull request by VVsssssk**
## Motivation
Support spconv2.0 in mmdet3d dev branch.  If the user installed spconv2.0, the code would use spconv2.0 first. And if spconv2.0 is not installed, the code will use mmcv's spconv op.

## BC-breaking (Optional)
.

## Checklist

- [x] fix get_start docs
- [x] fix ut
- [x] add spconv check in collect_env.py

### CODE REVIEW

File: mmdet3d/ops/sparse_block.py [hunk 1]
```diff
@@ -43,24 +48,44 @@ def __init__(self,
...
-        out.features = self.bn1(out.features)
-        out.features = self.relu(out.features)
+            if self.downsample is not None:
+                identity = self.downsample(x)
 
-        out = self.conv2(out)
-        out.features = self.bn2(out.features)
-        out.features = self.relu(out.features)
+            out = out.replace_feature(out.features + identity)
+            out = out.replace_feature(self.relu(out.features))
```
Comment (Tai-Wang) [inline mmdet3d/ops/sparse_block.py:69]: These two lines are the only differences between if-else? May simplify the codes

File: mmdet3d/ops/spconv/overwrite_spconv.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import itertools
+
+from mmcv.cnn.bricks.registry import CONV_LAYERS
+from torch.nn.parameter import Parameter
+
+
+def write_spconv():
```
Comment (Tai-Wang) [inline mmdet3d/ops/spconv/overwrite_spconv.py:8]: may rename as overwrite_spconv/write_spconv2/register_spconv2

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                dims = list(range(1, len(param.shape))) + [0]
+                param = param.permute(*dims)
+            destination[prefix + name] = param
+    for name, buf in self._buffers.items():
+        if buf is not None and name not in self._non_persistent_buffers_set:
+            destination[prefix + name] = buf if keep_vars else buf.detach()
+
+
+def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+                          missing_keys, unexpected_keys, error_msgs):
```
Comment (Tai-Wang) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:54]: Please add the corresponding docstring.

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        CONV_LAYERS._register_module(SubMConv2d, 'SubMConv2d', force=True)
+        CONV_LAYERS._register_module(SubMConv3d, 'SubMConv3d', force=True)
+        CONV_LAYERS._register_module(SubMConv4d, 'SubMConv4d', force=True)
+        SparseModule._load_from_state_dict = _load_from_state_dict
+        SparseModule._save_to_state_dict = _save_to_state_dict
+        return True
+
+
+def _save_to_state_dict(self, destination, prefix, keep_vars):
+    for name, param in self._parameters.items():
```
Comment (Tai-Wang) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:41]: Is there any difference for spconv 1.0 and 2.0 when saving and loading checkpoints? Maybe need some comments or a brief docstring to summarize the differences.

**Review by Tai-Wang — CHANGES_REQUESTED** [review]
Just need to fix minor comments.

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import itertools
+
+from mmcv.cnn.bricks.registry import CONV_LAYERS
+from torch.nn.parameter import Parameter
+
+
+def register_spconv2():
```
Comment (ZCMax) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:8]: Adding some docstring for explaining that Once spconv2.0 can be used, we need to register the spconv2.0 to overwrite the original spconv in mmcv.

File: docs/en/getting_started.md [hunk 6]
```diff
@@ -192,6 +192,15 @@ you can install it before installing MMCV.
 
 4. Some dependencies are optional. Simply running `pip install -v -e .` will only install the minimum runtime requirements. To use optional dependencies like `albumentations` and `imagecorruptions` either install them manually with `pip install -r requirements/optional.txt` or specify desired extras when calling `pip` (e.g. `pip install -v -e .[optional]`). Valid keys for the extras field are: `all`, `tests`, `build`, and `optional`.
 
+   We have supported spconv2.0. If you have installed spconv2.0, the code will use spconv2.0 first, and it has better performance. If necessary please refer to [spconv v2.x](https://github.com/traveller59/spconv) or use pip:
```
Comment (ZwwWayne) [inline docs/en/getting_started.md:195]: We have supported spconv2.0. If you have installed spconv2.0, the code will use spconv2.0 first, which could reduce the GPU memory. Users can use the following commands to install spconv2.0,
```bash
pip install cumm-cuxxx
pip install spconv-cuxxx
```
where `xxx` is the CUDA version in the environment.
For example, if you are using CUDA 10.2, the command will be `pip install cumm-cu102
 && pip install spconv-cu102`.
Supported CUDA versions include 10.2, 11.1, 11.3, 11.4. Users can also install it by building from source.
For more details please refer to [spconv v2.x](https://github.com/traveller59/spconv).

File: mmdet3d/models/middle_encoders/sparse_encoder.py [hunk 7]
```diff
@@ -1,11 +1,17 @@
 # Copyright (c) OpenMMLab. All rights reserved.
-from mmcv.ops import SparseConvTensor, SparseSequential
+
 from mmcv.runner import auto_fp16
 from torch import nn as nn
 
 from mmdet3d.ops import SparseBasicBlock, make_sparse_convmodule
+from mmdet3d.ops.spconv import spconv2_is_avalible
```
Comment (ZwwWayne) [inline mmdet3d/models/middle_encoders/sparse_encoder.py:7]: is_spconv2_available

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        elif strict:
+            missing_keys.append(key)
+
+    if strict:
+        for key, input_param in state_dict.items():
+            if key.startswith(prefix):
+                input_name = key[len(prefix):]
+                input_name = input_name.split(
+                    '.', 1)[0]  # get the name of param/buffer/child
+                if input_name not in self._modules\
```
Comment (ZwwWayne) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:104]:
```
if input_name not in self._modules \
    and input_name not in local_state:
```

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                continue
+
+            if isinstance(input_param, Parameter):
+                # backwards compatibility for serialized parameters
+                input_param = input_param.data
+            try:
+                param.copy_(input_param)
+            except Exception:
+                error_msgs.append(
+                    'While copying the parameter named "{}", '
```
Comment (ZwwWayne) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:91]: f-string

**Review by ZwwWayne — APPROVED** [review]

File: mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        CONV_LAYERS._register_module(SubMConv2d, 'SubMConv2d', force=True)
+        CONV_LAYERS._register_module(SubMConv3d, 'SubMConv3d', force=True)
+        CONV_LAYERS._register_module(SubMConv4d, 'SubMConv4d', force=True)
+        SparseModule._load_from_state_dict = _load_from_state_dict
+        SparseModule._save_to_state_dict = _save_to_state_dict
+        return True
+
+
+def _save_to_state_dict(self, destination, prefix, keep_vars):
+    """Rewrite this func to resolve spconv ops kernel weights incompatible.
```
Comment (ZwwWayne) [inline mmdet3d/ops/spconv/overwrite_spconv/write_spconv2.py:43]:
```python
"""Rewrite this func to compat the convolutional kernel weights between spconv 1.x in MMCV and 2.x in spconv2.x.

    Kernel weights in MMCV spconv has shape in (D,H,W,in_channel,out_channel), while those in spcon2.x
    is in (out_channel,D,H,W,in_channel).
"""
```

File: mmdet3d/utils/collect_env.py [hunk 11]
```diff
@@ -13,7 +14,7 @@ def collect_env():
     env_info['MMDetection'] = mmdet.__version__
     env_info['MMSegmentation'] = mmseg.__version__
     env_info['MMDetection3D'] = mmdet3d.__version__ + '+' + get_git_hash()[:7]
-
+    env_info['spconv2.0 is avalible'] = IS_SPCONV2_AVAILABLE
```
Comment (ZwwWayne) [inline mmdet3d/utils/collect_env.py:17]: spconv2.0 is avalible -> spconv2.0

**Review by ZwwWayne — APPROVED** [review]

**Comment by ammaryasirnaich** [comment]
@VVsssssk, I have one question after training with the spconv2. If we use the `mmdet3d/apis/inference.py` for doing `inference_detector(model, pcd)` than which version will it choose spconV2 or V1? However, I think it's using V1 which is causing a dimensional mismatch with the trained weights!
I will really appreciate if you can confirm this, thanks

**Pull request merged**